### PR TITLE
feat: allow access to private ECR repo in other regions than Greengrass Core region

### LINKED ISSUE #1048: (Nucleus): allow access to private ECR repo in other regions than Greengrass Core region
**Feature Description**
Allow access to private ECR repositories in other regions than the Greengrass Core region

**Use Case**
In order to avoid duplicating container images, one would like to be able to access ECR repositories in other regions than the one Greengrass Core device is registered to.

**Proposed Solution**
Derive the region to use for  `aws ecr get-login-password` from the `docker:` URI instead of using the region of the Greengrass Core device

**Other**

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by ChangxinDong**
**Issue #, if available:**
Resolves #1048

**Description of changes:**

**Why is this change necessary:**

**How was this change tested:**
- [ ] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/plugins/docker/DockerImageArtifactDownload01Test.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        NucleusPaths nucleusPaths = kernel.getNucleusPaths();
+        nucleusPaths.setComponentStorePath(tempRootDir);
+        ComponentStore store = new ComponentStore(nucleusPaths, platformResolver, recipeLoader);
+
+        EcrClient eastEcrClient = EcrClient.builder().httpClient(ProxyUtils.getSdkHttpClient())
+                .region(Region.of("us-east-1"))
+                .credentialsProvider(lazyCredentialProvider).build();
+        EcrClient westEcrClient = EcrClient.builder().httpClient(ProxyUtils.getSdkHttpClient())
+                .region(Region.of("us-west-1"))
+                .credentialsProvider(lazyCredentialProvider).build();
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/plugins/docker/DockerImageArtifactDownload01Test.java:88]: this is creating _real_ ecr clients. These can be mocks.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/plugins/docker/DockerImageArtifactDownload01Test.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    @Test
+    void GIVEN_component_with_ecr_private_image_as_artifact_WHEN_prepare_packages_THEN_download_image()
+            throws Exception {
+        ComponentIdentifier componentId =
+                new ComponentIdentifier("docker.image.from.ecr.private.registry01", new Semver("1.0.0"));
+        componentManager.preparePackages(Collections.singletonList(componentId)).get();
+
+        ArgumentCaptor argument = ArgumentCaptor.forClass(Registry.class);
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/plugins/docker/DockerImageArtifactDownload01Test.java:135]: `ArgumentCaptor<Registry>` should fix all the casting you are doing below here.

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessor.java [hunk 3]
```diff
@@ -57,25 +58,36 @@ public EcrAccessor(EcrClient client) {
         this.lazyCredentialProvider = null;
     }
 
-    private EcrClient getClient() {
+    /**
+     * Constructor for a client with region.
+     * @param actualRegion region
+     * @return EcrClient
+     */
+    public EcrClient getClient(String actualRegion) {
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessor.java:66]: Rename `actualRegion` -> `region`

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 4]
```diff
@@ -161,8 +162,11 @@ private File performDownloadSteps(Image image) throws PackageDownloadException,
                         // credentials for them on case by case basis
                         if (image.getRegistry().isEcrRegistry()) {
                             // Get auth token for ECR, which represents ECR registry credentials
+
+                            String actualRegion = getRegionFromArtifactUri(image.getArtifactUri().toString());
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:166]: What is `actualRegion`? it's clear in the context of this CR but won't be for someone who reads this code later. Let's find a better name for this that conveys this is the region where the image is hosted

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 5]
```diff
@@ -213,6 +217,27 @@ private File performDownloadSteps(Image image) throws PackageDownloadException,
...
+
+        String regionStr = "";
+
+        if (!Utils.isEmpty(artifactUriStr)) {
+            String[] arr = artifactUriStr.split("\\.");
+
+            if (arr.length > 1) {
+                for (int i = 0; i < arr.length; i++) {
+                    if ("amazonaws".equalsIgnoreCase(arr[i])) {
+                        regionStr = arr[i - 1];
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:231]: this array index may still be out of bounds.

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessor.java [hunk 6]
```diff
@@ -42,40 +42,37 @@ public class EcrAccessor {
     public EcrAccessor(DeviceConfiguration deviceConfiguration, LazyCredentialProvider lazyCredentialProvider) {
         this.deviceConfiguration = deviceConfiguration;
         this.lazyCredentialProvider = lazyCredentialProvider;
-        this.injectedClient = null;
     }
 
+
     /**
-     * Constructor for testing with a mocked client.
+     * Constructor for a client with region.
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessor.java:49]: this is not a constructor.

File: src/test/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessorTest.java [hunk 7]
```diff
@@ -29,19 +34,28 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/componentmanager/plugins/docker/EcrAccessorTest.java:37]: configure your IDE to not use `*` imports.

**Review by MikeDombo — APPROVED** [review]

**Review by indougnito — APPROVED** [review]

**Pull request merged**